Write working with user guide (#1100)
GetStream/stream-chat-swift: stream-chat-swift-docs/docs/guides/working-with-user.md

---
title: Working with User
---

> User object is accessed trough `Controller`. 
> There is special type of user object which is `CurrentUser`, representing the currently signed in user. 

Reviewer: ```suggestion
The User object is accessed through a `Controller`. 
There is a special type of user object which is `CurrentUser`, representing the currently signed-in user. 
```

```swift
/// User id of the user you want to work with
let userId = "yourUserId"

/// User controller for the intended user

> let userController = chatClient.usercController(userId: userId)

Reviewer: ```suggestion
let userController = chatClient.usersController(userId: userId)
```

/// Current user controller
let currentUserController = chatClient.currentUserController()
```

## Standard User

`CharUser` is a simple object, you can access the information and (un)mute and (un)flag the user.
```swift
let user = userController.user
print(user.name)
```

### Muting/unmuting user
```swift 
userController.mute()
userController.unmute()
```

> Both function have optional completion block to be called when the network request is finished

Reviewer: ```suggestion
Both functions have an optional completion block to be called when the network request is finished.
```

```swift 
userController.mute {

>     if let error = error {
>         print(error)
>     }

Reviewer 2: ```suggestion
    if let error = error {
        print(error)
        return
    }
```

    // Successfully finished
}
```

### Flagging/Unflagging user
```swift 
userController.flag()
userController.unflag()
```

> Both function have optional completion block to be called when the network request is finished

Reviewer: ```suggestion
Both functions have an optional completion block to be called when the network request is finished.
```

```swift 
userController.flag {

>     if let error = error {
>         print(error)
>     }

Reviewer 2: ```suggestion
    if let error = error {
        print(error)
        return
    }
```

    // Successfully finished
}
```

## Updating current user data

> You can call update user data function on current user controller

Reviewer: ```suggestion
You can call update user data function on the current user controller.
```

```swift 
currentUserController.updateUserData(name: "Luke Skywalker")
```

You can observe the changes in user data with delegate.

```swift 
// Don't forget to keep reference to the delegate. 
// `CurrentUserController` keeps only weak reference to the delegate.
let delegate = YourDelegate()
currentUserController.delegate = delegate

currentUserController.updateUserData(name: "Luke Skywalker")

class YourDelegate: CurrentChatUserControllerDelegate {
    func currentUserController(
        _ controller: _CurrentChatUserController<ExtraData>,
        didChangeCurrentUser: EntityChange<_CurrentChatUser<ExtraData>>
    ) {
        if case let .update(user) = didChangeCurrentUser {
            print(user.name) // Luke Skywalker
        }
    }
}
```

> You can use combine to observe the changes too

Reviewer: ```suggestion
You can use Combine to observe the changes too.
```

```swift
currentUserController
    .currentUserChangePublisher
    .compactMap({ (change: EntityChange<CurrentChatUser>) -> CurrentChatUser? in
        if case let .update(user) = change { return user } else { return nil }
    })
    .sink { print($0.name) } // Luke Skywalker
    .store(in: &cancellables)
```

## Observe unread count for user

To observe unread count on current user controller, you need to set a delegate. 

```swift
class YourDelegate: CurrentChatUserControllerDelegate {
    func currentUserController(
        _ controller: CurrentChatUserController, 
        didChangeCurrentUserUnreadCount count: UnreadCount
    ) {
        /// Handle the undread count
        UIApplication.shared.applicationIconBadgeNumber = count.messages
    }
}

// Don't forget to keep reference to the delegate. 
// `CurrentUserController` keeps only weak reference to the delegate.
let delegate = YourDelegate()
currentUserController.delegate = delegate
```

> You can use combine to observe unread count too

Reviewer: ```suggestion
You can use Combine to observe unread counts too.
```

```swift
currentUserController
    .unreadCountPublisher
    .map(\.messages)
    .sink { UIApplication.shared.applicationIconBadgeNumber = $0 }
    .store(in: &cancellables)
```

## Change user

To change the logged in user, simply change token provider of the chatClient.

```swift 
let token = ... // Obtain the chat token from your system
chatClient.tokenProvider = .static(token)

> ```

Reviewer 2: To change the current user we also have to initiate the reload via current user controller:
```swift
// Reload the current user
chatClient.currentUserController().reloadUserIfNeeded { error in
 ...
}
```

I'd also add some lines about `ChatClientConfig.shouldConnectAutomatically` flag, how it works, and how to manually control when client `connects/disconnects` from web-socket but that's the advanced topic 🤔 

@b-onc WDYT? Are we gonna have a dedicated page for `TokenProvider`, `reloadUserIfNeeded`, `connect/disconnect` stuff or we can briefly describe this in `current user` guide?

Reviewer 3: @evsaev good points, we'll have a dedicated Config page so i wouldn't put the flag here. `TokenProvider` and the rest belongs to Getting Started guide

Reviewer 3: `reloadUserIfNeeded` definitely belongs in this guide, but not the rest

---
title: Working with User
---

> The User object is accessed through a `Controller`. 

Reviewer 4: I think it makes sense to refer to a specific controller, i.e. `ChatUserController`

> There is a special type of user object which is `CurrentUser`, representing the currently signed-in user. 

Reviewer 4: Would be cool to showcase how some of `ChatUser` properties map to views in DemoApp. For instance name, imageURL, isOnline

Reviewer 3: also, needs explaining "why" we need this special entity

```swift
/// User id of the user you want to work with
let userId = "yourUserId"

/// User controller for the intended user

> let userController = chatClient.usersController(userId: userId)

Reviewer 4: ```suggestion
let userController = chatClient.userController(userId: userId)
```

/// Current user controller
let currentUserController = chatClient.currentUserController()
```

## Standard User

> `CharUser` is a simple object, you can access the information and (un)mute and (un)flag the user.

Reviewer 4: To introduce this entity smoother, we could slightly change the guide beginning to something like
"Stream Chat users are represented by `ChartUser` objects"

```suggestion
`ChartUser` is a simple object, you can access the information and (un)mute and (un)flag the user.
```

Author: did you mean `ChatUser`? 😅

```swift
let user = userController.user
print(user.name)
```

### Muting/unmuting user
```swift 
userController.mute()
userController.unmute()
```
Both functions have an optional completion block to be called when the network request is finished.
```swift 
userController.mute {
    if let error = error {
        print(error)
        return
    }
    // Successfully finished
}
```

### Flagging/Unflagging user
```swift 
userController.flag()
userController.unflag()
```
Both functions have an optional completion block to be called when the network request is finished.
```swift 
userController.flag {
    if let error = error {
        print(error)
        return
    }
    // Successfully finished
}
```

## Updating current user data

> You can call update user data function on the current user controller.

Reviewer 4: I don't know how this works, so what I thought reading this might be similar to what customer would think.
And I thought something like: "In which cases I want to update user's data? Do I have to call this every time before accessing a user model for it to be fresh?"

```swift 
currentUserController.updateUserData(name: "Luke Skywalker")
```

> You can observe the changes in user data with delegate.

Reviewer 3: It needs combine and swiftUI showcasing too

Author: `UserController` has nothing to do with (Swift)UI

Reviewer 3: https://github.com/GetStream/stream-chat-swift/blob/main/Sources/StreamChat/Controllers/UserController/UserController+SwiftUI.swift

Author: There is nothing UI related in the referenced file, the `SwiftUI.ObservableObject` is just internal typealias to `Combine.ObservableObject`. The file should be merged with `UserController+Combine.swift`, the same applies for `CurrentUserController`

```swift 
// Don't forget to keep reference to the delegate. 

> // `CurrentUserController` keeps only weak reference to the delegate.

Reviewer 3: Let's remove these

Author: I know I struggled with this when I was a newbie dev (my first year), I spend more than I would like on debugging issues like this 😅 
What is the benefit of removing those?

Reviewer 3: We don't want to let guides dive too deep into development basics and divert from the actual point. Our delegates work like any other delegate, we should only document what we're doing different

let delegate = YourDelegate()
currentUserController.delegate = delegate

> currentUserController.updateUserData(name: "Luke Skywalker")

Reviewer 3: we should use more real-world scenarios for these, eg in a ViewController

class YourDelegate: CurrentChatUserControllerDelegate {
    func currentUserController(
        _ controller: _CurrentChatUserController<ExtraData>,
        didChangeCurrentUser: EntityChange<_CurrentChatUser<ExtraData>>
    ) {
        if case let .update(user) = didChangeCurrentUser {
            print(user.name) // Luke Skywalker
        }
    }
}
```

> You can use Combine to observe the changes too.

Reviewer 3: Having these as 3rd level headers makes navigation easier

Author: I'm already using 3rd level for different scenarios above. Should I use the 4th level for delegation/combine code sample titles?

Reviewer 3: works

```swift
currentUserController
    .currentUserChangePublisher
    .compactMap({ (change: EntityChange<CurrentChatUser>) -> CurrentChatUser? in
        if case let .update(user) = change { return user } else { return nil }
    })
    .sink { print($0.name) } // Luke Skywalker
    .store(in: &cancellables)
```

## Observe unread count for user

To observe unread count on current user controller, you need to set a delegate. 

```swift
class YourDelegate: CurrentChatUserControllerDelegate {
    func currentUserController(
        _ controller: CurrentChatUserController, 
        didChangeCurrentUserUnreadCount count: UnreadCount
    ) {
        /// Handle the undread count
        UIApplication.shared.applicationIconBadgeNumber = count.messages
    }
}

// Don't forget to keep reference to the delegate. 
// `CurrentUserController` keeps only weak reference to the delegate.
let delegate = YourDelegate()
currentUserController.delegate = delegate
```

You can use Combine to observe unread counts too.
```swift
currentUserController
    .unreadCountPublisher
    .map(\.messages)
    .sink { UIApplication.shared.applicationIconBadgeNumber = $0 }
    .store(in: &cancellables)
```

> ## Change user

Reviewer 4: It's a super minor really, but headers are not consistent in a sense that some of them use infinitives (updating) and some present simple (change). They appear in navigation and if we came up with unified way of doing this it might slightly improve overall aftertaste for the docs

Author: What is the time used in the other guides? So I can match it.

Reviewer 3: We use infinitives, so it should be "Changing User"
